lib/ansible/module_utils/csharp/Ansible.Basic.cs

```diff
@@ -275,14 +276,31 @@ public void Debug(string message)
 
         public void Deprecate(string message, string version)
         {
-            deprecations.Add(new Dictionary<string, string>() { { "msg", message }, { "version", version } });
+            deprecations.Add(new Dictionary<string, string>() {
+                { "msg", message }, { "version", version }, { "collection_name", null } });
+            LogEvent(String.Format("[DEPRECATION WARNING] {0} {1}", message, version));
+        }
+
+        public void Deprecate(string message, string version, string collectionName)
+        {
+            deprecations.Add(new Dictionary<string, string>() {
+                { "msg", message }, { "version", version }, { "collection_name", collectionName } });
             LogEvent(String.Format("[DEPRECATION WARNING] {0} {1}", message, version));
         }
 
         public void Deprecate(string message, DateTime date)
         {
             string isoDate = date.ToString("yyyy-MM-dd");
-            deprecations.Add(new Dictionary<string, string>() { { "msg", message }, { "date", isoDate } });
+            deprecations.Add(new Dictionary<string, string>() {
+                { "msg", message }, { "date", isoDate }, { "collection_name", null } });
+            LogEvent(String.Format("[DEPRECATION WARNING] {0} {1}", message, isoDate));
+        }
+
+        public void Deprecate(string message, DateTime date, string collectionName)
```

Comment:
Instead of having a copy of `public void Deprecate(string message, DateTime date)` you can either

* Have the first function call the second one with it defining `null` for the `collectionName` param

```csharp
public void Deprecate(string message, DateTime date)
{
    Deprecate(message, date, null)
}

public void Deprecate(string message, DateTime date, string collectionName)
{
    ...
}
```

* Or set `collectionName` as an option parameter with a default value like

```csharp
public void Deprecate(string message, DateTime date, string collectionName = null)
{
    ...
}
```

The latter is simpler as you don't need to define the function twice but it does have the problem where `$null` passed in from PowerShell it becomes an empty string but it really depends on what is consuming this data if `None == ""`.

Reply:
Actually I was using `public void Deprecate(string message, DateTime date, string collectionName = null)` first, but there's a bug in Windows 2012 which causes this to fail compiling (if the other `Deprecate` overload for versions is using a default parameter for collectionName too).

Reply:
That's definitely weird, let's just go for option 1 then.

Reply:
Implemented.